example/src/main/java/org/apache/rocketmq/example/benchmark/TransactionProducer.java

```diff
@@ -175,84 +226,171 @@ public static Options buildCommandlineOptions(final Options options) {
-        if (ischeck) {
-            return LocalTransactionState.UNKNOW;
-        }
-        return LocalTransactionState.COMMIT_MESSAGE;
+    public LocalTransactionState executeLocalTransaction(Message msg, Object arg) {
+        return parseFromMsg(msg).sendResult;
     }
-}
 
-class TransactionCheckListenerBImpl implements TransactionCheckListener {
-    private boolean ischeckffalse;
-    private StatsBenchmarkTProducer statsBenchmarkTProducer;
-
-    public TransactionCheckListenerBImpl(boolean ischeckffalse,
-        StatsBenchmarkTProducer statsBenchmarkTProducer) {
-        this.ischeckffalse = ischeckffalse;
-        this.statsBenchmarkTProducer = statsBenchmarkTProducer;
+    private MsgMeta parseFromMsg(Message msg) {
+        byte[] bs = msg.getBody();
+        ByteBuffer buf = ByteBuffer.wrap(bs);
+        MsgMeta msgMeta = new MsgMeta();
+        msgMeta.batchId = buf.getLong();
+        msgMeta.msgId = buf.getLong();
+        msgMeta.sendResult = LocalTransactionState.values()[buf.get()];
+        msgMeta.checkResult = new ArrayList<>();
+        for (int i = 0; i < TransactionProducer.MAX_CHECK_RESULT_IN_MSG; i++) {
+            msgMeta.checkResult.add(LocalTransactionState.values()[buf.get()]);
+        }
+        return msgMeta;
     }
 
     @Override
-    public LocalTransactionState checkLocalTransactionState(MessageExt msg) {
-        statsBenchmarkTProducer.getCheckRequestSuccessCount().incrementAndGet();
-        if (ischeckffalse) {
+    public LocalTransactionState checkLocalTransaction(MessageExt msg) {
+        MsgMeta msgMeta = parseFromMsg(msg);
+        if (msgMeta.batchId != sendConfig.batchId) {
+            // message not generated in this test
+            return LocalTransactionState.ROLLBACK_MESSAGE;
+        }
+        statBenchmark.getCheckCount().incrementAndGet();
 
+        int times = 0;
+        try {
+            String checkTimes = msg.getUserProperty(MessageConst.PROPERTY_TRANSACTION_CHECK_TIMES);
+            times = Integer.parseInt(checkTimes);
+        } catch (Exception e) {
             return LocalTransactionState.ROLLBACK_MESSAGE;
         }
+        times = times <= 0 ? 1 : times;
+
+        boolean dup;
+        synchronized (cache) {
+            Integer oldCheckLog = cache.get(msgMeta.msgId);
+            Integer newCheckLog;
+            if (oldCheckLog == null) {
+                newCheckLog = 1 << (times - 1);
+            } else {
+                newCheckLog = oldCheckLog | (1 << (times - 1));
+            }
+            dup = newCheckLog.equals(oldCheckLog);
+        }
+        if (dup) {
+            statBenchmark.getDuplicatedCheckCount().incrementAndGet();
+        }
+        if (msgMeta.sendResult != LocalTransactionState.UNKNOW) {
+            System.out.printf("%s unexpected check: msgId=%s,txId=%s,checkTimes=%s,sendResult=%s\n",
+                    new SimpleDateFormat("HH:mm:ss,SSS").format(new Date()),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The SimpleDateFormat object was created without setting the time zone. Make sure that you want to use the default time zone or use letter z, Z or X in the pattern. Otherwise, if you do not call setTimeZone() on the created SimpleDateFormat object, you might get an unexpected date and time when using the object.